Source/AwsCommonRuntimeKit/sdkutils/endpoint/EndpointsRuleEngine.swift

```diff
@@ -0,0 +1,61 @@
+//  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//  SPDX-License-Identifier: Apache-2.0.
+
+import AwsCSdkUtils
+
+/// Rule engine for matching endpoint rules
+public class CRTAWSEndpointsRuleEngine {
+
+    let rawValue: OpaquePointer
+    let allocator: Allocator
+
+    /// Initialize a new rule engine
+    /// - Parameters:
+    ///   - partitions: JSON string containing partition data
+    ///   - ruleSet: The rule set string to use for the rule engine
+    ///   - allocator: The allocator to use for creating rule engine
+    public init(partitions: String, ruleSet: String, allocator: Allocator = defaultAllocator) throws {
+        let getRawValue: () throws -> OpaquePointer = {
+            guard let partitions = (partitions.withByteCursor { partitionsCursor in
+                aws_partitions_config_new_from_string(allocator.rawValue, partitionsCursor)
+            }) else {
+                throw CommonRunTimeError.crtError(.makeFromLastError())
+            }
+
+            guard let ruleSet = (ruleSet.withByteCursor { ruleSetCursor in
+                aws_endpoints_ruleset_new_from_string(allocator.rawValue, ruleSetCursor)
+            }) else {
+                throw CommonRunTimeError.crtError(.makeFromLastError())
+            }
+
+            defer {
+                aws_partitions_config_release(partitions)
```

Comment:
possible memory leak: if `aws_endpoints_ruleset_new_from_string()` fails, then `partitions` will leak

we need to declare the deferred release of each thin immediately after its successfully created

Reply:
Thanks, fixed.